pkg/liveshare/test/server.go

```diff
@@ -134,6 +149,13 @@ func (s *Server) Err() <-chan error {
 	return s.errCh
 }
 
+func (s *Server) WriteToObjectStream(obj interface{}) error {
+	if s.objectStream == nil {
+		return errors.New("object stream not set")
+	}
+	return s.objectStream.WriteObject(obj)
```

Comment:
It does, hence the nil check. When we wrote it we knew it wasn't totally safe but we felt it was good enough for testing.

Reply:
The nil check isn't sufficient to ensure correct behavior. Consider this program with two threads, A and B:
```
A:
  global = &T{x: 1}
B:
  if global != nil { print(global.x) }
```
It is not guaranteed that thread B prints 1. It could print 0 on a weakly ordered architectures such as M1 because the hardware reorders the t.x=1 and global=t assignments. This could make the program behave differently on different machines. Also, the race detector will fail for this program; by putting known data races into the code, you undermine the value of the race detector for detecting unknown data races.

I don't think it's good practice to write sloppy code just because the program is a test. Like any program, a test has a job to do---to save maintainers time by detecting regressions---and when it is buggy it fails to do that job. In other words, rather than saving time, it will waste it.

Reply:
Yeah, I meant that the current test code does not make use of multiple threads to execute this code path. I agree with you that we should still not introduce sloppy code. The nil check plus known usage patterns felt good enough but perhaps we should revisit.